// mineral-ui/Button/Button.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.componentTheme = undefined;

var _jsx = function () { var REACT_ELEMENT_TYPE = typeof Symbol === "function" && Symbol.for && Symbol.for("react.element") || 0xeac7; return function createRawReactElement(type, props, key, children) { var defaultProps = type && type.defaultProps; var childrenLength = arguments.length - 3; if (!props && childrenLength !== 0) { props = {}; } if (props && defaultProps) { for (var propName in defaultProps) { if (props[propName] === void 0) { props[propName] = defaultProps[propName]; } } } else if (!props) { props = defaultProps || {}; } if (childrenLength === 1) { props.children = children; } else if (childrenLength > 1) { var childArray = Array(childrenLength); for (var i = 0; i < childrenLength; i++) { childArray[i] = arguments[i + 3]; } props.children = childArray; } return { $$typeof: REACT_ELEMENT_TYPE, type: type, key: key === undefined ? null : '' + key, ref: null, props: props, _owner: null }; }; }();

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; /**
                                                                                                                                                                                                                                                                   * Copyright 2017 CA
                                                                                                                                                                                                                                                                   *
                                                                                                                                                                                                                                                                   * Licensed under the Apache License, Version 2.0 (the "License");
                                                                                                                                                                                                                                                                   * you may not use this file except in compliance with the License.
                                                                                                                                                                                                                                                                   * You may obtain a copy of the License at
                                                                                                                                                                                                                                                                   *
                                                                                                                                                                                                                                                                   *    http://www.apache.org/licenses/LICENSE-2.0
                                                                                                                                                                                                                                                                   *
                                                                                                                                                                                                                                                                   * Unless required by applicable law or agreed to in writing, software
                                                                                                                                                                                                                                                                   * distributed under the License is distributed on an "AS IS" BASIS,
                                                                                                                                                                                                                                                                   * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
                                                                                                                                                                                                                                                                   * See the License for the specific language governing permissions and
                                                                                                                                                                                                                                                                   * limitations under the License.
                                                                                                                                                                                                                                                                   */

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _styles = require('../styles');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var componentTheme = exports.componentTheme = function componentTheme(baseTheme) {
  return _extends({
    Button_backgroundColor: baseTheme.color_gray_20,
    Button_backgroundColor_active: baseTheme.color_gray_30,
    Button_backgroundColor_focus: baseTheme.color_gray_20,
    Button_backgroundColor_hover: baseTheme.color_gray_10,
    Button_backgroundColor_minimal_active: baseTheme.color_gray_20,
    Button_backgroundColor_minimal_hover: baseTheme.color_gray_10,
    Button_backgroundColor_primary: baseTheme.color_theme_60,
    Button_backgroundColor_primary_active: baseTheme.color_theme_70,
    Button_backgroundColor_primary_focus: baseTheme.color_theme_60,
    Button_backgroundColor_primary_hover: baseTheme.color_theme_50,
    Button_borderColor: baseTheme.borderColor,
    Button_borderColor_focus: baseTheme.color_white,
    Button_borderRadius: baseTheme.borderRadius_1,
    Button_borderWidth: 1, // px
    Button_boxShadow_focus: '0 0 0 1px ' + baseTheme.borderColor_focus,
    Button_color_text: baseTheme.color_gray_100,
    Button_color_text_minimal: baseTheme.color_text_primary,
    Button_color_text_primary: baseTheme.color_text_onprimary,
    Button_fontWeight: baseTheme.fontWeight_semiBold,
    Button_paddingHorizontal: baseTheme.space_inset_sm,
    Button_paddingIconOnly_small: (0, _styles.pxToEm)(3),
    Button_paddingIconOnly_medium: (0, _styles.pxToEm)(7),
    Button_paddingIconOnly_large: (0, _styles.pxToEm)(7),
    Button_paddingIconOnly_jumbo: (0, _styles.pxToEm)(13),
    Button_height_small: baseTheme.size_small,
    Button_height_medium: baseTheme.size_medium,
    Button_height_large: baseTheme.size_large,
    Button_height_jumbo: baseTheme.size_jumbo,

    ButtonContent_fontSize: baseTheme.fontSize_ui,
    ButtonContent_fontSize_small: (0, _styles.pxToEm)(12),

    ButtonIcon_margin: baseTheme.space_inset_sm

  }, baseTheme);
};

function chooseColor(_ref, theme) {
  var disabled = _ref.disabled,
      primary = _ref.primary,
      minimal = _ref.minimal;

  if (disabled) {
    return theme.color_text_disabled;
  } else if (primary) {
    return theme.Button_color_text_primary;
  } else if (minimal) {
    return theme.Button_color_text_minimal;
  } else {
    return theme.Button_color_text;
  }
}

var styles = {
  button: function button(props) {
    var theme = componentTheme(props.theme);
    var circular = props.circular,
        disabled = props.disabled,
        fullWidth = props.fullWidth,
        minimal = props.minimal,
        primary = props.primary,
        size = props.size,
        text = props.text,
        variant = props.variant;


    if (variant !== 'regular') {
      // prettier-ignore
      theme = _extends({}, theme, {
        Button_backgroundColor_primary: theme['backgroundColor_' + variant],
        Button_backgroundColor_primary_active: theme['backgroundColor_' + variant + '_active'],
        Button_backgroundColor_primary_focus: theme['backgroundColor_' + variant + '_focus'],
        Button_backgroundColor_primary_hover: theme['backgroundColor_' + variant + '_hover'],
        Button_boxShadow_focus: '0 0 0 1px ' + theme['borderColor_' + variant + '_focus'],
        Button_color_text: theme['color_text_' + variant],
        Button_color_text_minimal: theme['color_text_' + variant],
        Button_color_text_primary: theme['color_text_on' + variant]
      });
    }

    var color = chooseColor(props, theme);
    return {
      backgroundColor: function () {
        if (disabled && !minimal) {
          return theme.color_gray_30;
        } else if (primary) {
          return theme.Button_backgroundColor_primary;
        } else if (minimal) {
          return 'transparent';
        } else {
          return theme.Button_backgroundColor;
        }
      }(),
      borderColor: disabled || primary || minimal ? 'transparent' : theme.Button_borderColor,
      borderRadius: circular ? parseFloat(theme['Button_height_' + size]) / 2 + 'em' : theme.Button_borderRadius,
      borderStyle: 'solid',
      borderWidth: theme.Button_borderWidth + 'px',
      color: color,
      cursor: disabled ? 'default' : 'pointer',
      display: 'inline-block',
      fontWeight: theme.Button_fontWeight,
      height: theme['Button_height_' + size],
      // if the user puts in a small icon in a large button
      // we want to force the button to be round/square
      // (really just pertinent on icon-only buttons)
      minWidth: theme['Button_height_' + size],
      padding: text === undefined ? theme['Button_paddingIconOnly_' + size] : '0 ' + theme.Button_paddingHorizontal,
      textDecoration: 'none',
      verticalAlign: 'middle',
      width: fullWidth && '100%',
      '&:focus': {
        backgroundColor: function () {
          if (primary) {
            return theme.Button_backgroundColor_primary_focus;
          } else if (minimal) {
            return theme.Button_backgroundColor_minimal_focus;
          } else {
            return theme.Button_backgroundColor_focus;
          }
        }(),
        borderColor: theme.Button_borderColor_focus,
        boxShadow: theme.Button_boxShadow_focus,
        color: color,
        textDecoration: 'none'
      },
      '&:hover': {
        backgroundColor: function () {
          if (!disabled) {
            if (primary) {
              return theme.Button_backgroundColor_primary_hover;
            } else if (minimal) {
              return theme.Button_backgroundColor_minimal_hover;
            } else {
              return theme.Button_backgroundColor_hover;
            }
          }
        }(),
        color: color,
        textDecoration: 'none'
      },
      // `:active` must be last, to follow LVHFA order:
      // https://developer.mozilla.org/en-US/docs/Web/CSS/:active
      '&:active': {
        backgroundColor: function () {
          if (!disabled) {
            if (primary) {
              return theme.Button_backgroundColor_primary_active;
            } else if (minimal) {
              return theme.Button_backgroundColor_minimal_active;
            } else {
              return theme.Button_backgroundColor_active;
            }
          }
        }(),
        color: color
      },
      '&::-moz-focus-inner': { border: 0 },

      '& [role="img"]': {
        boxSizing: 'content-box',
        fill: disabled || primary || minimal || variant !== 'regular' ? 'currentColor' : theme.Button_backgroundColor_primary,
        display: 'block',

        '&:first-child': {
          marginLeft: theme.direction === 'rtl' ? theme.ButtonIcon_margin : null,
          marginRight: theme.direction === 'ltr' ? theme.ButtonIcon_margin : null
        },

        '&:last-child': {
          marginLeft: theme.direction === 'ltr' ? theme.ButtonIcon_margin : null,
          marginRight: theme.direction === 'rtl' ? theme.ButtonIcon_margin : null
        },

        '&:only-child': {
          margin: 0
        }
      }
    };
  },
  content: function content(props) {
    var theme = componentTheme(props.theme);
    var size = props.size;


    var paddings = void 0;

    var fontSize = size === 'small' ? theme.ButtonContent_fontSize_small : theme.ButtonContent_fontSize;

    if (size === undefined || size === 'large' || size === 'jumbo') {
      var padding = (0, _styles.getNormalizedValue)(theme.Button_paddingHorizontal, fontSize);
      paddings = {
        '&:first-child': {
          paddingLeft: theme.direction === 'ltr' ? padding : null,
          paddingRight: theme.direction === 'rtl' ? padding : null
        },

        '&:last-child': {
          paddingLeft: theme.direction === 'rtl' ? padding : null,
          paddingRight: theme.direction === 'ltr' ? padding : null
        }
      };
    }

    return _extends({
      display: 'inline-block',
      maxWidth: '100%',
      overflow: 'hidden',
      textOverflow: 'ellipsis',
      whiteSpace: 'nowrap',
      wordWrap: 'normal'
    }, {

      display: 'block',
      fontSize: fontSize,
      lineHeight: (0, _styles.getNormalizedValue)(theme['Button_height_' + size], fontSize)
    }, paddings);
  },
  inner: {
    alignItems: 'center',
    display: 'inline-flex',
    justifyContent: 'center',
    maxHeight: '100%',
    width: '100%'
  }
};

var iconSize = {
  small: 'medium',
  medium: 'medium',
  large: (0, _styles.pxToEm)(24),
  jumbo: (0, _styles.pxToEm)(24)
};

var Content = (0, _styles.createStyledComponent)('span', styles.content);
var Inner = (0, _styles.createStyledComponent)('span', styles.inner);

function isTypeButton(type) {
  return ['button', 'submit', 'reset'].indexOf(type) !== -1;
}

function filterProps(_ref2) {
  var element = _ref2.element,
      type = _ref2.type;

  // When element is a component, e.g. ReactRouterLink,
  // these are not filtered automatically by rootEl
  var invalidComponentProps = ['primary', 'text', 'variant'];
  var invalidLinkProps = element === 'a' && isTypeButton(type) ? ['type'] : [];
  return Array.prototype.concat(invalidComponentProps, invalidLinkProps);
}

// Button's root node must be created outside of render, so that the entire DOM
// element is replaced only when the element prop is changed, otherwise it is
// updated in place
function createRootNode(props) {
  var _props$element = props.element,
      element = _props$element === undefined ? Button.defaultProps.element : _props$element;


  return (0, _styles.createStyledComponent)(element, styles.button, {
    filterProps: filterProps(props),
    includeStyleReset: true,
    rootEl: element
  });
}

/**
 * The Button component represents a clickable button.
 * Buttons draw attention to actions that can be performed in your app.
 * Buttons are used to trigger any sort of event.
 */

var Button = function (_Component) {
  _inherits(Button, _Component);

  function Button() {
    var _ref3;

    var _temp, _this, _ret;

    _classCallCheck(this, Button);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref3 = Button.__proto__ || Object.getPrototypeOf(Button)).call.apply(_ref3, [this].concat(args))), _this), _this.rootNode = createRootNode(_this.props), _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Button, [{
    key: 'componentWillUpdate',
    value: function componentWillUpdate(nextProps) {
      if (this.props.element !== nextProps.element) {
        this.rootNode = createRootNode(nextProps);
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          iconStart = _props.iconStart,
          iconEnd = _props.iconEnd,
          _props$size = _props.size,
          size = _props$size === undefined ? Button.defaultProps.size : _props$size,
          _props$type = _props.type,
          type = _props$type === undefined ? Button.defaultProps.type : _props$type,
          _props$variant = _props.variant,
          variant = _props$variant === undefined ? Button.defaultProps.variant : _props$variant,
          restProps = _objectWithoutProperties(_props, ['children', 'iconStart', 'iconEnd', 'size', 'type', 'variant']);

      var rootProps = _extends({
        size: size,
        text: children,
        type: type,
        variant: variant
      }, restProps);

      var Root = this.rootNode;

      var startIcon = iconStart ? (0, _react.cloneElement)(iconStart, { size: iconSize[size], key: 'iconStart' }) : null;
      var endIcon = iconEnd ? (0, _react.cloneElement)(iconEnd, { size: iconSize[size], key: 'iconEnd' }) : null;

      return _react2.default.createElement(
        Root,
        rootProps,
        _jsx(Inner, {}, void 0, startIcon, children && _jsx(Content, {
          size: size
        }, void 0, children), endIcon)
      );
    }
  }]);

  return Button;
}(_react.Component);

Button.propTypes = {
  /** Rendered content of the component */
  children: typeof React$Node === 'function' ? require('prop-types').instanceOf(React$Node) : require('prop-types').any,

  /** Displays a circular Button */
  circular: require('prop-types').bool,

  /** Disables the Button */
  disabled: require('prop-types').bool,

  /** Element to be used as the root node - e.g. `a` can be used to create a link that is styled like a Button */
  element: typeof $FlowFixMe === 'function' ? require('prop-types').instanceOf($FlowFixMe) : require('prop-types').any,
  // Should allow string | React class
  /** Stretch Button to fill its container */
  fullWidth: require('prop-types').bool,

  /** Icon that goes after the children*/
  iconEnd: typeof React$Element === 'function' ? require('prop-types').instanceOf(React$Element) : require('prop-types').any,

  /** Icon that goes before the children */
  iconStart: typeof React$Element === 'function' ? require('prop-types').instanceOf(React$Element) : require('prop-types').any,

  /** Display a minimal Button */
  minimal: require('prop-types').bool,

  /** Called with the click event */
  onClick: require('prop-types').func,

  /** Display a primary Button */
  primary: require('prop-types').bool,

  /** Available sizes */
  size: require('prop-types').oneOf(['small', 'medium', 'large', 'jumbo']),

  /** Available types */
  type: require('prop-types').string,

  /** Available variants */
  variant: require('prop-types').oneOf(['regular', 'danger', 'success', 'warning'])
};
Button.defaultProps = {
  element: 'button',
  size: 'large',
  type: 'button',
  variant: 'regular'
};
exports.default = Button;
