// mineral-ui/Checkbox/CheckboxGroup.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.componentTheme = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; /**
                                                                                                                                                                                                                                                                   * Copyright 2017 CA
                                                                                                                                                                                                                                                                   *
                                                                                                                                                                                                                                                                   * Licensed under the Apache License, Version 2.0 (the "License");
                                                                                                                                                                                                                                                                   * you may not use this file except in compliance with the License.
                                                                                                                                                                                                                                                                   * You may obtain a copy of the License at
                                                                                                                                                                                                                                                                   *
                                                                                                                                                                                                                                                                   *    http://www.apache.org/licenses/LICENSE-2.0
                                                                                                                                                                                                                                                                   *
                                                                                                                                                                                                                                                                   * Unless required by applicable law or agreed to in writing, software
                                                                                                                                                                                                                                                                   * distributed under the License is distributed on an "AS IS" BASIS,
                                                                                                                                                                                                                                                                   * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
                                                                                                                                                                                                                                                                   * See the License for the specific language governing permissions and
                                                                                                                                                                                                                                                                   * limitations under the License.
                                                                                                                                                                                                                                                                   */

exports.default = CheckboxGroup;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _themes = require('../themes');

var _ChoiceGroup = require('../Choice/ChoiceGroup');

var _ChoiceGroup2 = _interopRequireDefault(_ChoiceGroup);

var _Checkbox = require('./Checkbox');

var _Checkbox2 = _interopRequireDefault(_Checkbox);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var componentTheme = exports.componentTheme = function componentTheme(baseTheme) {
  return _extends({}, (0, _themes.mapComponentThemes)({
    name: 'ChoiceGroup',
    theme: (0, _ChoiceGroup.componentTheme)(baseTheme)
  }, {
    name: 'CheckboxGroup',
    theme: {}
  }, baseTheme));
};

var Root = (0, _themes.createThemedComponent)(_ChoiceGroup2.default, function (_ref) {
  var baseTheme = _ref.theme;

  return _extends({}, (0, _themes.mapComponentThemes)({
    name: 'CheckboxGroup',
    theme: componentTheme(baseTheme)
  }, {
    name: 'ChoiceGroup',
    theme: {}
  }, baseTheme));
});

/**
 * CheckboxGroup allows authors to construct a group of
 * [Checkboxes](../checkbox) and provides a simpler API than working with
 * Checkbox directly.
 *
 * CheckboxGroup allows users to select multiple options from a list.
 */
function CheckboxGroup(_ref2) {
  var otherRootProps = _ref2.rootProps,
      restProps = _objectWithoutProperties(_ref2, ['rootProps']);

  var rootProps = _extends({
    rootProps: _extends({}, otherRootProps),
    input: _Checkbox2.default,
    type: 'checkbox'
  }, restProps);

  return _react2.default.createElement(Root, rootProps);
}
