// mineral-ui/es/Card/CardTitleMenu.js
var _jsx = function () { var REACT_ELEMENT_TYPE = typeof Symbol === "function" && Symbol.for && Symbol.for("react.element") || 0xeac7; return function createRawReactElement(type, props, key, children) { var defaultProps = type && type.defaultProps; var childrenLength = arguments.length - 3; if (!props && childrenLength !== 0) { props = {}; } if (props && defaultProps) { for (var propName in defaultProps) { if (props[propName] === void 0) { props[propName] = defaultProps[propName]; } } } else if (!props) { props = defaultProps || {}; } if (childrenLength === 1) { props.children = children; } else if (childrenLength > 1) { var childArray = Array(childrenLength); for (var i = 0; i < childrenLength; i++) { childArray[i] = arguments[i + 3]; } props.children = childArray; } return { $$typeof: REACT_ELEMENT_TYPE, type: type, key: key === undefined ? null : '' + key, ref: null, props: props, _owner: null }; }; }();

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

/**
 * Copyright 2017 CA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { createStyledComponent, pxToEm } from '../styles';
import Button from '../Button';
import Dropdown from '../Dropdown';
import IconMoreHoriz from '../Icon/IconMoreHoriz';

/*
 * A large Button, even with zero'd padding, is still a bit too large in this
 * context. These styles allow the Button to shrink, but the Icon remains the
 * same size.
 */
var MenuButton = createStyledComponent(Button, function (_ref) {
  var theme = _ref.theme;
  return {
    height: 'auto',
    minWidth: 0,
    overflow: 'hidden',
    padding: 0,

    // Inner
    '& > span': {
      display: 'block',
      margin: '-' + pxToEm(2)
    },

    '& [role="img"]': {
      fill: theme.color_text
    }
  };
});

/**
 * CardTitleMenu
 */
export default function CardTitleMenu(_ref2) {
  var data = _ref2.data,
      _ref2$triggerTitle = _ref2.triggerTitle,
      triggerTitle = _ref2$triggerTitle === undefined ? 'Card actions' : _ref2$triggerTitle,
      restProps = _objectWithoutProperties(_ref2, ['data', 'triggerTitle']);

  var rootProps = _extends({
    data: data,
    placement: 'bottom-end'
  }, restProps);
  var buttonProps = {
    iconStart: _jsx(IconMoreHoriz, {
      title: triggerTitle
    }),
    minimal: true
  };

  return React.createElement(
    Dropdown,
    rootProps,
    React.createElement(MenuButton, buttonProps)
  );
}
