// mineral-ui/Link/Link.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.componentTheme = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; /**
                                                                                                                                                                                                                                                                   * Copyright 2017 CA
                                                                                                                                                                                                                                                                   *
                                                                                                                                                                                                                                                                   * Licensed under the Apache License, Version 2.0 (the "License");
                                                                                                                                                                                                                                                                   * you may not use this file except in compliance with the License.
                                                                                                                                                                                                                                                                   * You may obtain a copy of the License at
                                                                                                                                                                                                                                                                   *
                                                                                                                                                                                                                                                                   *    http://www.apache.org/licenses/LICENSE-2.0
                                                                                                                                                                                                                                                                   *
                                                                                                                                                                                                                                                                   * Unless required by applicable law or agreed to in writing, software
                                                                                                                                                                                                                                                                   * distributed under the License is distributed on an "AS IS" BASIS,
                                                                                                                                                                                                                                                                   * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
                                                                                                                                                                                                                                                                   * See the License for the specific language governing permissions and
                                                                                                                                                                                                                                                                   * limitations under the License.
                                                                                                                                                                                                                                                                   */

exports.default = Link;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _styles = require('../styles');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var componentTheme = exports.componentTheme = function componentTheme(baseTheme) {
  return _extends({
    Link_borderColor_focus: baseTheme.borderColor_focus,
    Link_color: baseTheme.color_text_primary,
    Link_color_active: baseTheme.color_text_primary_active,
    Link_color_hover: baseTheme.color_text_primary_hover,
    Link_color_focus: baseTheme.color_text_primary_focus

  }, baseTheme);
};

var linkStyles = function linkStyles(_ref) {
  var variant = _ref.variant,
      baseTheme = _ref.theme;

  var theme = componentTheme(baseTheme);

  if (variant !== 'regular') {
    // prettier-ignore
    theme = _extends({}, theme, {
      Link_borderColor_focus: baseTheme['borderColor_' + variant + '_focus'],
      Link_color: baseTheme['color_text_' + variant],
      Link_color_hover: baseTheme['color_text_' + variant + '_hover'],
      Link_color_focus: baseTheme['color_text_' + variant + '_focus'],
      Link_color_active: baseTheme['color_text_' + variant + '_active']
    });
  }

  return {
    color: theme.Link_color,
    textDecoration: 'none',

    '&:hover': {
      color: theme.Link_color_hover,
      textDecoration: 'underline'
    },
    '&:focus': {
      color: theme.Link_color_focus,
      outline: '1px solid ' + theme.Link_borderColor_focus,
      outlineOffset: '2px'
    },
    // `:active` must be last, to follow LVHFA order:
    // https://developer.mozilla.org/en-US/docs/Web/CSS/:active
    '&:active': {
      color: theme.Link_color_active
    }
  };
};

/**
 * The Link component creates a hyperlink to external pages, files, anchors on the same page, or another URL.
 */
function Link(_ref2) {
  var children = _ref2.children,
      _ref2$element = _ref2.element,
      element = _ref2$element === undefined ? 'a' : _ref2$element,
      _ref2$variant = _ref2.variant,
      variant = _ref2$variant === undefined ? 'regular' : _ref2$variant,
      restProps = _objectWithoutProperties(_ref2, ['children', 'element', 'variant']);

  var rootProps = _extends({
    variant: variant
  }, restProps);
  var Root = (0, _styles.createStyledComponent)(element, linkStyles, {
    displayName: 'Link',
    filterProps: ['variant']
  });

  return _react2.default.createElement(
    Root,
    rootProps,
    children
  );
}
