// mineral-ui/es/Box/Box.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

/**
 * Copyright 2017 CA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { Component } from 'react';
import { createStyledComponent, getResponsiveStyles } from '../styles';

var spacingTypes = {
  marginHorizontal: 'inline',
  marginStart: 'inline',
  marginEnd: 'inline',
  marginLeft: 'inline',
  marginRight: 'inline',
  paddingHorizontal: 'inline',
  paddingStart: 'inline',
  paddingEnd: 'inline',
  paddingLeft: 'inline',
  paddingRight: 'inline',
  padding: 'inset',
  margin: 'stack',
  marginVertical: 'stack',
  marginTop: 'stack',
  marginBottom: 'stack',
  paddingVertical: 'stack',
  paddingTop: 'stack',
  paddingBottom: 'stack'
};

var getMeasurement = function getMeasurement(value) {
  return typeof value === 'number' && value < 1 ? value * 100 + '%' : value;
};

var getSpaceValue = function getSpaceValue(property, theme, value) {
  return theme['space_' + spacingTypes[property] + '_' + value] || getMeasurement(value);
};

var getSpacingStyles = function getSpacingStyles(property, restProps, rtl) {
  // prettier-ignore
  var spacingPropKeys = ['', 'Horizontal', 'Vertical', 'Start', 'End', 'Bottom', 'Left', 'Right', 'Top'].map(function (suffix) {
    return '' + property + suffix;
  }).filter(function (key) {
    return restProps[key] !== undefined;
  });

  var setStyles = function setStyles(directions, key, styles) {
    directions.forEach(function (direction) {
      styles['' + property + direction] = restProps[key];
    });
  };

  return spacingPropKeys.reduce(function (acc, key) {
    if (key === property) {
      setStyles(['Bottom', 'Left', 'Right', 'Top'], key, acc);
    } else if (key.indexOf('Horizontal') !== -1) {
      setStyles(['Left', 'Right'], key, acc);
    } else if (key.indexOf('Vertical') !== -1) {
      setStyles(['Bottom', 'Top'], key, acc);
    } else if (key.indexOf('Start') !== -1) {
      setStyles(['' + (rtl ? 'Right' : 'Left')], key, acc);
    } else if (key.indexOf('End') !== -1) {
      setStyles(['' + (rtl ? 'Left' : 'Right')], key, acc);
    } else {
      acc[key] = restProps[key];
    }
    return acc;
  }, {});
};

var styles = {
  root: function root(_ref) {
    var breakpoints = _ref.breakpoints,
        height = _ref.height,
        inline = _ref.inline,
        theme = _ref.theme,
        width = _ref.width,
        restProps = _objectWithoutProperties(_ref, ['breakpoints', 'height', 'inline', 'theme', 'width']);

    var rtl = theme.direction === 'rtl';

    var mapValueToProperty = function mapValueToProperty(property, value) {
      var map = _extends({
        display: function display(value) {
          return value ? 'inline-block' : 'block';
        },
        height: getMeasurement,
        width: getMeasurement
      }, ['margin', 'padding'].reduce(function (acc, property) {
        Object.keys(getSpacingStyles(property, restProps, rtl)).forEach(function (style) {
          acc[style] = function (value) {
            return getSpaceValue(property, theme, value);
          };
        });
        return acc;
      }, {}));

      return map[property](value);
    };

    return getResponsiveStyles({
      breakpoints: breakpoints,
      mapValueToProperty: mapValueToProperty,
      styles: _extends({
        display: inline,
        height: height
      }, getSpacingStyles('margin', restProps, rtl), getSpacingStyles('padding', restProps, rtl), {
        width: width
      }),
      theme: theme
    });
  }
};

// Box's root node must be created outside of render, so that the entire DOM
// element is replaced only when the element prop is changed, otherwise it is
// updated in place
function createRootNode(props) {
  var _props$element = props.element,
      element = _props$element === undefined ? Box.defaultProps.element : _props$element;


  return createStyledComponent(element, styles.root, {
    includeStyleReset: true,
    rootEl: element
  });
}

/**
 * Box component provides an easy way to apply standardized size & space to your
 * layout.
 *
 * Box is the basis of the other layout components: [Flex](../flex),
 * [FlexItem](../flex-item), and [StartEnd](../start-end).
 */

var Box = function (_Component) {
  _inherits(Box, _Component);

  function Box() {
    var _ref2;

    var _temp, _this, _ret;

    _classCallCheck(this, Box);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref2 = Box.__proto__ || Object.getPrototypeOf(Box)).call.apply(_ref2, [this].concat(args))), _this), _this.rootNode = createRootNode(_this.props), _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Box, [{
    key: 'componentWillUpdate',
    value: function componentWillUpdate(nextProps) {
      if (this.props.element !== nextProps.element) {
        this.rootNode = createRootNode(nextProps);
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var Root = this.rootNode;

      return React.createElement(Root, this.props);
    }
  }]);

  return Box;
}(Component);

Box.propTypes = {
  /**
   * Media query (min-width) breakpoints along which to apply props marked
   * "&#xfeff;[[Responsive-capable]](#responsive)&#xfeff;"
   */
  breakpoints: require('prop-types').arrayOf(require('prop-types').oneOfType([require('prop-types').number, require('prop-types').string])),

  /** Rendered root HTML element */
  element: require('prop-types').string,

  /** Sets the box height. [[Responsive-capable]](#responsive) */
  height: require('prop-types').oneOfType([require('prop-types').number, require('prop-types').string, require('prop-types').arrayOf(require('prop-types').oneOfType([require('prop-types').number, require('prop-types').string]))]),

  /** Renders Box as an inline-block [[Responsive-capable]](#responsive) */
  inline: require('prop-types').oneOfType([require('prop-types').bool, require('prop-types').arrayOf(require('prop-types').bool)]),

  /** Applies a margin on all sides [[Responsive-capable]](#responsive) */
  margin: require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string, require('prop-types').arrayOf(require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string]))]),

  /** Applies a bottom margin [[Responsive-capable]](#responsive) */
  marginBottom: require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string, require('prop-types').arrayOf(require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string]))]),

  /**
   * Applies a right margin when the language is left-to-right and left margin
   * [for RTL languages](#rtl) [[Responsive-capable]](#responsive)
   */
  marginEnd: require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string, require('prop-types').arrayOf(require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string]))]),

  /** Applies left & right margins [[Responsive-capable]](#responsive) */
  marginHorizontal: require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string, require('prop-types').arrayOf(require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string]))]),

  /** Applies a left margin [[Responsive-capable]](#responsive) */
  marginLeft: require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string, require('prop-types').arrayOf(require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string]))]),

  /** Applies a right margin [[Responsive-capable]](#responsive) */
  marginRight: require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string, require('prop-types').arrayOf(require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string]))]),

  /**
   * Applies a left margin when the language is left-to-right and right margin
   * [for RTL languages](#rtl) [[Responsive-capable]](#responsive)
   */
  marginStart: require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string, require('prop-types').arrayOf(require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string]))]),

  /** Applies a top margin [[Responsive-capable]](#responsive) */
  marginTop: require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string, require('prop-types').arrayOf(require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string]))]),

  /**
   * Applies top & bottom margins [[Responsive-capable]](#responsive)
   */
  marginVertical: require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string, require('prop-types').arrayOf(require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string]))]),

  /** Applies padding to all sides [[Responsive-capable]](#responsive) */
  padding: require('prop-types').oneOfType([require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').number, require('prop-types').string, require('prop-types').arrayOf(require('prop-types').oneOfType([require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').number, require('prop-types').string]))]),

  /** Applies bottom padding [[Responsive-capable]](#responsive) */
  paddingBottom: require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string, require('prop-types').arrayOf(require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string]))]),

  /**
   * Applies right padding when the language is left-to-right and left padding
   * [for RTL languages](#rtl) [[Responsive-capable]](#responsive)
   */
  padddingEnd: require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string, require('prop-types').arrayOf(require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string]))]),

  /** Applies left & right padding [[Responsive-capable]](#responsive) */
  paddingHorizontal: require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string, require('prop-types').arrayOf(require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string]))]),

  /** Applies left padding [[Responsive-capable]](#responsive) */
  paddingLeft: require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string, require('prop-types').arrayOf(require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string]))]),

  /** Applies right padding [[Responsive-capable]](#responsive) */
  paddingRight: require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string, require('prop-types').arrayOf(require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string]))]),

  /**
   * Applies left padding when the language is left-to-right and right padding
   * [for RTL languages](#rtl) [[Responsive-capable]](#responsive)
   */
  padddingStart: require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string, require('prop-types').arrayOf(require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string]))]),

  /** Applies bottom padding [[Responsive-capable]](#responsive) */
  paddingTop: require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string, require('prop-types').arrayOf(require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string]))]),

  /** Applies top & bottom margins [[Responsive-capable]](#responsive) */
  paddingVertical: require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string, require('prop-types').arrayOf(require('prop-types').oneOfType([require('prop-types').oneOf(['xxs']), require('prop-types').oneOf(['xs']), require('prop-types').oneOf(['sm']), require('prop-types').oneOf(['md']), require('prop-types').oneOf(['lg']), require('prop-types').oneOf(['xl']), require('prop-types').oneOf(['xxl']), require('prop-types').number, require('prop-types').string]))]),

  /** Sets the box width [[Responsive-capable]](#responsive) */
  width: require('prop-types').oneOfType([require('prop-types').number, require('prop-types').string, require('prop-types').arrayOf(require('prop-types').oneOfType([require('prop-types').number, require('prop-types').string]))])
};
Box.defaultProps = {
  element: 'div'
};
export default Box;
