// mineral-ui/colors/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _palette = require('./palette');

Object.defineProperty(exports, 'default', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_palette).default;
  }
});

var _blue = require('./blue');

Object.defineProperty(exports, 'blue', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_blue).default;
  }
});

var _common = require('./common');

Object.defineProperty(exports, 'common', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_common).default;
  }
});

var _dusk = require('./dusk');

Object.defineProperty(exports, 'dusk', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_dusk).default;
  }
});

var _gray = require('./gray');

Object.defineProperty(exports, 'gray', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_gray).default;
  }
});

var _green = require('./green');

Object.defineProperty(exports, 'green', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_green).default;
  }
});

var _indigo = require('./indigo');

Object.defineProperty(exports, 'indigo', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_indigo).default;
  }
});

var _lime = require('./lime');

Object.defineProperty(exports, 'lime', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_lime).default;
  }
});

var _magenta = require('./magenta');

Object.defineProperty(exports, 'magenta', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_magenta).default;
  }
});

var _orange = require('./orange');

Object.defineProperty(exports, 'orange', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_orange).default;
  }
});

var _purple = require('./purple');

Object.defineProperty(exports, 'purple', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_purple).default;
  }
});

var _red = require('./red');

Object.defineProperty(exports, 'red', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_red).default;
  }
});

var _sky = require('./sky');

Object.defineProperty(exports, 'sky', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_sky).default;
  }
});

var _slate = require('./slate');

Object.defineProperty(exports, 'slate', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_slate).default;
  }
});

var _teal = require('./teal');

Object.defineProperty(exports, 'teal', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_teal).default;
  }
});

var _yellow = require('./yellow');

Object.defineProperty(exports, 'yellow', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_yellow).default;
  }
});

var _getColor = require('./getColor');

Object.defineProperty(exports, 'getColor', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_getColor).default;
  }
});

var _getReadableTextColor = require('./getReadableTextColor');

Object.defineProperty(exports, 'getReadableTextColor', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_getReadableTextColor).default;
  }
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
