// mineral-ui/colors/keyedColors.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _blue = require('./blue');

var _blue2 = _interopRequireDefault(_blue);

var _dusk = require('./dusk');

var _dusk2 = _interopRequireDefault(_dusk);

var _gray = require('./gray');

var _gray2 = _interopRequireDefault(_gray);

var _green = require('./green');

var _green2 = _interopRequireDefault(_green);

var _indigo = require('./indigo');

var _indigo2 = _interopRequireDefault(_indigo);

var _lime = require('./lime');

var _lime2 = _interopRequireDefault(_lime);

var _magenta = require('./magenta');

var _magenta2 = _interopRequireDefault(_magenta);

var _orange = require('./orange');

var _orange2 = _interopRequireDefault(_orange);

var _purple = require('./purple');

var _purple2 = _interopRequireDefault(_purple);

var _red = require('./red');

var _red2 = _interopRequireDefault(_red);

var _sky = require('./sky');

var _sky2 = _interopRequireDefault(_sky);

var _slate = require('./slate');

var _slate2 = _interopRequireDefault(_slate);

var _teal = require('./teal');

var _teal2 = _interopRequireDefault(_teal);

var _yellow = require('./yellow');

var _yellow2 = _interopRequireDefault(_yellow);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Copyright 2017 CA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

exports.default = {
  blue: _blue2.default,
  dusk: _dusk2.default,
  gray: _gray2.default,
  green: _green2.default,
  indigo: _indigo2.default,
  lime: _lime2.default,
  magenta: _magenta2.default,
  orange: _orange2.default,
  purple: _purple2.default,
  red: _red2.default,
  sky: _sky2.default,
  slate: _slate2.default,
  teal: _teal2.default,
  yellow: _yellow2.default
};
