// mineral-ui/themes/createTheme.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; /**
                                                                                                                                                                                                                                                                   * Copyright 2017 CA
                                                                                                                                                                                                                                                                   *
                                                                                                                                                                                                                                                                   * Licensed under the Apache License, Version 2.0 (the "License");
                                                                                                                                                                                                                                                                   * you may not use this file except in compliance with the License.
                                                                                                                                                                                                                                                                   * You may obtain a copy of the License at
                                                                                                                                                                                                                                                                   *
                                                                                                                                                                                                                                                                   *    http://www.apache.org/licenses/LICENSE-2.0
                                                                                                                                                                                                                                                                   *
                                                                                                                                                                                                                                                                   * Unless required by applicable law or agreed to in writing, software
                                                                                                                                                                                                                                                                   * distributed under the License is distributed on an "AS IS" BASIS,
                                                                                                                                                                                                                                                                   * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
                                                                                                                                                                                                                                                                   * See the License for the specific language governing permissions and
                                                                                                                                                                                                                                                                   * limitations under the License.
                                                                                                                                                                                                                                                                   */

exports.default = createTheme;

var _colors = require('../colors');

var _colors2 = _interopRequireDefault(_colors);

var _createColorRamp = require('./createColorRamp');

var _createColorRamp2 = _interopRequireDefault(_createColorRamp);

var _fontSizeBase = require('./fontSizeBase');

var _fontSizeBase2 = _interopRequireDefault(_fontSizeBase);

var _pxToEm = require('../styles/pxToEm');

var _pxToEm2 = _interopRequireDefault(_pxToEm);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var spacing_quarter = (0, _pxToEm2.default)(2);
var spacing_half = (0, _pxToEm2.default)(4);
var spacing_1x = (0, _pxToEm2.default)(8);
var spacing_2x = (0, _pxToEm2.default)(16);
var spacing_3x = (0, _pxToEm2.default)(24);
var spacing_4x = (0, _pxToEm2.default)(32);
var spacing_8x = (0, _pxToEm2.default)(64);

var grays = (0, _createColorRamp2.default)('gray', 'color_gray', _colors2.default);

function createTheme() {
  var baseColor = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'blue';
  var overrides = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

  /* Theme key naming convention:
   *
   *   [property]_[target]_[variant]_[state]
   *
   * E.g., A key for the border-color on focused "success" buttons and inputs
   * would look like the following.
   *
   *   borderColor_success_focus
   *
   * Only [property] is required. Irrelevant parts are skipped.
   */

  var primaries = (0, _createColorRamp2.default)(baseColor, 'color_theme', _colors2.default);

  // prettier-ignore
  return _extends({
    backgroundColor_disabled: _colors2.default.gray_20,
    backgroundColor_danger: _colors2.default.red_60,
    backgroundColor_danger_muted: _colors2.default.red_20,
    backgroundColor_danger_active: _colors2.default.red_70,
    backgroundColor_danger_activeMuted: _colors2.default.red_30,
    backgroundColor_danger_focus: _colors2.default.red_60,
    backgroundColor_danger_hover: _colors2.default.red_50,
    backgroundColor_success: _colors2.default.green_80,
    backgroundColor_success_muted: _colors2.default.green_20,
    backgroundColor_success_active: _colors2.default.green_90,
    backgroundColor_success_activeMuted: _colors2.default.green_30,
    backgroundColor_success_focus: _colors2.default.green_80,
    backgroundColor_success_hover: _colors2.default.green_70,
    backgroundColor_warning: _colors2.default.orange_60,
    backgroundColor_warning_muted: _colors2.default.orange_20,
    backgroundColor_warning_active: _colors2.default.orange_70,
    backgroundColor_warning_activeMuted: _colors2.default.orange_30,
    backgroundColor_warning_focus: _colors2.default.orange_60,
    backgroundColor_warning_hover: _colors2.default.orange_50,

    backgroundColor_input: _colors2.default.white,

    backgroundColor_link_focus: _colors2.default.gray_30,

    borderColor: _colors2.default.gray_40,
    borderColor_active: primaries.color_theme_60,
    borderColor_focus: primaries.color_theme_60,
    borderColor_hover: primaries.color_theme_60,
    borderColor_danger: _colors2.default.red_60,
    borderColor_danger_active: _colors2.default.red_70,
    borderColor_danger_focus: _colors2.default.red_70,
    borderColor_danger_hover: _colors2.default.red_70,
    borderColor_danger_muted: _colors2.default.red_40,
    borderColor_success: _colors2.default.green_70,
    borderColor_success_active: _colors2.default.green_80,
    borderColor_success_focus: _colors2.default.green_80,
    borderColor_success_hover: _colors2.default.green_80,
    borderColor_success_muted: _colors2.default.green_40,
    borderColor_warning: _colors2.default.orange_60,
    borderColor_warning_active: _colors2.default.orange_70,
    borderColor_warning_focus: _colors2.default.orange_70,
    borderColor_warning_hover: _colors2.default.orange_70,
    borderColor_warning_muted: _colors2.default.orange_40,

    borderRadius_1: (0, _pxToEm2.default)(3),

    breakpoint_narrow: 512,
    breakpoint_medium: 768,
    breakpoint_wide: 1024,

    color_black: _colors2.default.black,

    color_caption: _colors2.default.gray_80

  }, grays, {

    color_placeholder: _colors2.default.gray_60

  }, primaries, {

    color_text: _colors2.default.gray_100,
    color_text_disabled: _colors2.default.gray_50,
    color_text_danger: _colors2.default.red_60,
    color_text_danger_active: _colors2.default.red_70,
    color_text_danger_focus: _colors2.default.red_60,
    color_text_danger_hover: _colors2.default.red_50,
    color_text_ondanger: _colors2.default.white,
    color_text_primary: primaries.color_theme_60,
    color_text_primary_active: primaries.color_theme_70,
    color_text_primary_focus: primaries.color_theme_60,
    color_text_primary_hover: primaries.color_theme_50,
    color_text_onprimary: _colors2.default.white,
    color_text_success: _colors2.default.green_90,
    color_text_success_active: _colors2.default.green_100,
    color_text_success_focus: _colors2.default.green_90,
    color_text_success_hover: _colors2.default.green_80,
    color_text_onsuccess: _colors2.default.white,
    color_text_warning: _colors2.default.orange_70,
    color_text_warning_active: _colors2.default.orange_80,
    color_text_warning_focus: _colors2.default.orange_70,
    color_text_warning_hover: _colors2.default.orange_60,
    color_text_onwarning: _colors2.default.white,

    color_white: _colors2.default.white,

    direction: 'ltr',

    fontFamily: 'Open Sans',
    fontFamily_system: '-apple-system, system-ui, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol"',
    fontFamily_monospace: '"SF Mono", "Droid Sans Mono", "Source Code Pro", Monaco, Consolas, "Courier New", Courier, monospace',

    fontSize_base: _fontSizeBase2.default,
    fontSize_h1: (0, _pxToEm2.default)(34),
    fontSize_h2: (0, _pxToEm2.default)(28),
    fontSize_h3: (0, _pxToEm2.default)(22),
    fontSize_h4: (0, _pxToEm2.default)(18),
    fontSize_h5: (0, _pxToEm2.default)(14),
    fontSize_h6: (0, _pxToEm2.default)(14),
    fontSize_mouse: (0, _pxToEm2.default)(11),
    fontSize_prose: (0, _pxToEm2.default)(16),
    fontSize_ui: (0, _pxToEm2.default)(14),

    fontWeight_regular: 400,
    fontWeight_semiBold: 600,
    fontWeight_bold: 700,
    fontWeight_extraBold: 800,

    lineHeight: 1.25,
    lineHeight_prose: 1.5,

    shadow_1: '0 1px 2px 0 rgba(0,0,0,0.2), 0 2px 4px 0 rgba(0,0,0,0.2)',
    shadow_2: '0 2px 4px 0 rgba(0,0,0,0.2), 0 4px 8px 0 rgba(0,0,0,0.2)',
    shadow_3: '0 4px 8px 0 rgba(0,0,0,0.2), 0 8px 16px 0 rgba(0,0,0,0.2)',
    shadow_4: '0 8px 16px 0 rgba(0,0,0,0.2), 0 20px 16px -8px rgba(0,0,0,0.2)',
    shadow_5: '0 16px 24px 0 rgba(0,0,0,0.2), 0 32px 24px -16px rgba(0,0,0,0.2)',

    size_small: (0, _pxToEm2.default)(24),
    size_medium: (0, _pxToEm2.default)(32),
    size_large: (0, _pxToEm2.default)(40),
    size_jumbo: (0, _pxToEm2.default)(52),

    space_inline_xxs: spacing_quarter,
    space_inline_xs: spacing_half,
    space_inline_sm: spacing_1x,
    space_inline_md: spacing_2x,
    space_inline_lg: spacing_3x,
    space_inline_xl: spacing_4x,
    space_inline_xxl: spacing_8x,

    space_inset_sm: spacing_1x,
    space_inset_md: spacing_2x,
    space_inset_lg: spacing_3x,

    space_stack_xxs: spacing_quarter,
    space_stack_xs: spacing_half,
    space_stack_sm: spacing_1x,
    space_stack_md: spacing_2x,
    space_stack_lg: spacing_3x,
    space_stack_xl: spacing_4x,
    space_stack_xxl: spacing_8x,

    zIndex_100: 100,
    zIndex_200: 200,
    zIndex_400: 400,
    zIndex_800: 800,
    zIndex_1600: 1600

  }, overrides);
}
