// mineral-ui/StartEnd/StartEnd.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _jsx = function () { var REACT_ELEMENT_TYPE = typeof Symbol === "function" && Symbol.for && Symbol.for("react.element") || 0xeac7; return function createRawReactElement(type, props, key, children) { var defaultProps = type && type.defaultProps; var childrenLength = arguments.length - 3; if (!props && childrenLength !== 0) { props = {}; } if (props && defaultProps) { for (var propName in defaultProps) { if (props[propName] === void 0) { props[propName] = defaultProps[propName]; } } } else if (!props) { props = defaultProps || {}; } if (childrenLength === 1) { props.children = children; } else if (childrenLength > 1) { var childArray = Array(childrenLength); for (var i = 0; i < childrenLength; i++) { childArray[i] = arguments[i + 3]; } props.children = childArray; } return { $$typeof: REACT_ELEMENT_TYPE, type: type, key: key === undefined ? null : '' + key, ref: null, props: props, _owner: null }; }; }();

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = StartEnd;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Flex = require('../Flex');

var _Flex2 = _interopRequireDefault(_Flex);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; } /**
                                                                                                                                                                                                                              * Copyright 2017 CA
                                                                                                                                                                                                                              *
                                                                                                                                                                                                                              * Licensed under the Apache License, Version 2.0 (the "License");
                                                                                                                                                                                                                              * you may not use this file except in compliance with the License.
                                                                                                                                                                                                                              * You may obtain a copy of the License at
                                                                                                                                                                                                                              *
                                                                                                                                                                                                                              *    http://www.apache.org/licenses/LICENSE-2.0
                                                                                                                                                                                                                              *
                                                                                                                                                                                                                              * Unless required by applicable law or agreed to in writing, software
                                                                                                                                                                                                                              * distributed under the License is distributed on an "AS IS" BASIS,
                                                                                                                                                                                                                              * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
                                                                                                                                                                                                                              * See the License for the specific language governing permissions and
                                                                                                                                                                                                                              * limitations under the License.
                                                                                                                                                                                                                              */

/**
 * StartEnd provides a simple way to align components to the start and end of a
 * container.
 */
function StartEnd(props) {
  var children = props.children,
      ignoreJustifyContent = props.justifyContent,
      priority = props.priority,
      ignoreWrap = props.wrap,
      restProps = _objectWithoutProperties(props, ['children', 'justifyContent', 'priority', 'wrap']);

  var rootProps = _extends({
    justifyContent: 'between'
  }, restProps);
  var flexItems = void 0;
  if (_react.Children.count(children) === 2) {
    var growFactors = {
      start: [1, 0],
      end: [0, 1],
      both: [1, 1],
      none: [undefined, undefined]
    };
    flexItems = _react.Children.map(children, function (child, index) {
      return index ? _jsx(_Flex.FlexItem, {
        grow: growFactors[priority || 'none'][1]
      }, void 0, child) : _jsx(_Flex.FlexItem, {
        grow: growFactors[priority || 'none'][0]
      }, void 0, child);
    });
  } else {
    throw new Error('StartEnd must have exactly two children.');
  }

  // $FlowFixMe - `direction` can't have the '-reverse' options from Flex
  return _react2.default.createElement(
    _Flex2.default,
    rootProps,
    flexItems
  );
}
StartEnd.propTypes = {
  /**
   * Media query (min-width) breakpoints along which to apply props marked
   * "&#xfeff;[[Responsive-capable]](#responsive)&#xfeff;"
   */
  breakpoints: require('prop-types').arrayOf(require('prop-types').oneOfType([require('prop-types').number, require('prop-types').string])),

  /** Must be exactly two nodes */
  children: typeof React$Node === 'function' ? require('prop-types').instanceOf(React$Node).isRequired : require('prop-types').any.isRequired,

  /**
   * Direction of flow of items along the main axis
   * [[Responsive-capable]](#responsive)
   */
  direction: require('prop-types').oneOfType([require('prop-types').oneOf(['column']), require('prop-types').oneOf(['row']), require('prop-types').arrayOf(require('prop-types').oneOf(['column', 'row']))]),

  /** @Private From Flex; filtered out */
  justifyContent: require('prop-types').any,

  /**
   * Displays the second, rather than the first, child with a majority of the
   * available width
   */
  priority: require('prop-types').oneOf(['start', 'end', 'both', false]),

  /** @Private From Flex; filtered out */
  wrap: require('prop-types').any
};
